nes-network/bindings/network/lib.rs

```diff
@@ -0,0 +1,359 @@
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+use nes_network::protocol::{ConnectionIdentifier, ThisConnectionIdentifier, TupleBuffer};
+use nes_network::receiver::{ReceiverChannel, ReceiverChannelResult};
+use nes_network::sender::{SenderChannel, TrySendDataResult};
+use nes_network::*;
+use once_cell::sync;
+use std::collections::HashMap;
+use std::collections::hash_map::Entry;
+use std::error::Error;
+use tracing::warn;
+use std::pin::Pin;
+use std::str::FromStr;
+use std::sync::{Arc, Mutex};
+
+#[cxx::bridge]
+pub mod ffi {
+    enum SendResult {
+        Ok,
+        Closed,
+        Full,
+    }
+
+    struct SerializedTupleBuffer {
+        sequence_number: usize,
+        origin_id: usize,
+        chunk_number: usize,
+        number_of_tuples: usize,
+        watermark: usize,
+        last_chunk: bool,
+    }
+
+    unsafe extern "C++" {
+        include!("NetworkBindings.hpp");
+        type TupleBufferBuilder;
+        #[allow(non_snake_case)]
+        fn setMetadata(self: Pin<&mut TupleBufferBuilder>, meta: &SerializedTupleBuffer);
+        #[allow(non_snake_case)]
+        fn setData(self: Pin<&mut TupleBufferBuilder>, data: &[u8]);
+        #[allow(non_snake_case)]
+        fn addChildBuffer(self: Pin<&mut TupleBufferBuilder>, data: &[u8]);
+        #[allow(non_snake_case)]
+        fn identifyThread(thread_name: &str, worker_id: &str);
+    }
+
+    extern "Rust" {
+        type ReceiverNetworkService;
+        type SenderNetworkService;
+        type SenderDataChannel;
+        type ReceiverDataChannel;
+
+        fn enable_memcom() -> Result<()>;
+        fn init_receiver_service(connection_addr: String, worker_id: String) -> Result<()>;
+        fn receiver_instance(connection_addr: String) -> Result<Box<ReceiverNetworkService>>;
+        fn init_sender_service(connection_addr: String, worker_id: String) -> Result<()>;
+        fn sender_instance(connection_addr: String) -> Result<Box<SenderNetworkService>>;
+
+        fn register_receiver_channel(
+            server: &mut ReceiverNetworkService,
+            channel_identifier: String,
+        ) -> Box<ReceiverDataChannel>;
+        fn receive_buffer(
+            receiver_channel: &ReceiverDataChannel,
+            builder: Pin<&mut TupleBufferBuilder>,
+        ) -> Result<bool>;
+        fn interrupt_receive(receiver_channel: &ReceiverDataChannel);
+
+        fn close_receiver_channel(channel: Box<ReceiverDataChannel>);
+
+        fn register_sender_channel(
+            server: &SenderNetworkService,
+            connection_identifier: String,
+            channel_identifier: String,
+        ) -> Result<Box<SenderDataChannel>>;
+
+        fn close_sender_channel(channel: Box<SenderDataChannel>);
+        fn flush_channel(channel: &SenderDataChannel) -> bool;
+        fn try_send_on_channel(
+            channel: &SenderDataChannel,
+            metadata: SerializedTupleBuffer,
```

Comment:
At this point it is a `SerializedHeader`, because the data has not been attached yet?